[Chapter 13 - MicroProfile Reactive Messaging 3.0] Adding new chapter for MP Reactive Messaging 3.0 (#69)
microprofile/microprofile-tutorial: modules/ROOT/pages/chapter13/chapter13.adoc

> = MicroProfile Reactive Messaging

Reviewer: The nav does not include this chapter. BTW, what's chapter 12?

Author: Nav has been updated in another PR. 
Chapter 12 is about MicroProfile GraphQL.

{reactive-messaging-spec-name} provides an annotation-driven programming model for building asynchronous, message-driven applications with support for backpressure. In practice, applications connect to brokers such as Kafka or AMQP-based systems by using runtime-specific connectors.

Using this model, developers can build scalable and resilient microservices that communicate through events rather than tightly coupled synchronous calls.

== Topics Covered

* Event-driven architecture (EDA) overview
* Reactive Streams fundamentals
* Adding dependencies (Maven, Gradle, and connector dependencies)
* Defining messaging logic with CDI beans
* Leveraging CDI features in messaging beans
* Channels and wiring
* Declarative messaging with `@Incoming` and `@Outgoing`
* Supported method signatures and message processing patterns
* Programmatic messaging with `@Channel` and `@Emitter`
* Managing backpressure in reactive messaging
* Integration with other MicroProfile specifications
* Reactive stream operations

> * Best practices and production patterns

Reviewer: Where is this covered?

For learning, local development, and testing, many runtimes provide an internal in-memory connector. This connector allows messages to flow between channels within the same application, without requiring an external broker. For production systems, use a messaging provider such as Apache Kafka or RabbitMQ.

=== Maven Dependency

Add the following dependency to your `pom.xml`:

[source,xml]
----
<dependency>
    <groupId>org.eclipse.microprofile.reactive.messaging</groupId>
    <artifactId>microprofile-reactive-messaging-api</artifactId>
    <version>${reactive-messaging-spec-version}</version>           
    <scope>provided</scope>
</dependency>
----

This dependency provides the MicroProfile Reactive Messaging API. The `provided` scope indicates that the implementation is supplied by your MicroProfile-compatible runtime or application server rather than bundled inside your application.

=== Gradle Dependency

Add the following dependency to your `build.gradle`:

[source,groovy]
----
dependencies {

>     compileOnly "org.eclipse.microprofile.reactive.messaging:microprofile-reactive-messaging-api:3.0.1"

Reviewer: Should it use the attribute instead of 3.0.1?

Author: Updated with attribute.

* It performs a simple transformation by updating the order status to `PAID`.
* It sends the updated order to the `payment-authorized` channel for downstream processing by services such as Inventory or Order status handlers.

== Leveraging CDI Features in Messaging Beans

Since messaging beans are CDI-managed, you can leverage the full power of CDI features such as dependency injection, scopes, interceptors, and events. This allows you to build complex message-processing logic while keeping your code clean, maintainable, and easy to test.

=== Dependency Injection in Messaging Beans

You can inject other CDI beans, services, or resources into your messaging beans to perform tasks such as persistence, external API calls, notifications, or business-rule validation. For example, an `OrderStatusEventHandler` bean can inject an `OrderService` to update the order in the database and a `NotificationService` to send payment confirmations:

[source,java]
----
// ... 

@ApplicationScoped
public class OrderStatusEventHandler {

    @Inject
    OrderService orderService;

    @Inject
    NotificationService notificationService;

    @Incoming("payment-authorized")

>     public void onPaymentAuthorized(Long orderId) {

Reviewer: The producer returns an "Order", here you consume an `orderId`. Something is wrong.

Author: Fixed.

* the `shoppingcart` component submits an order to the existing `order` service through REST
* the `order` service persists the order and publishes an `OrderCreated` event to the `order-created` channel
* the `payment` service consumes that event and can publish a payment result such as `payment-authorized`
* the `inventory` and `shipment` services can also then react to these events independently
* additional processors can filter, enrich, audit, or route the same event stream without changing the public REST API

*Channel Types*:

* *Internal Channels*: Connect components within the same application when methods are wired together by channel name.
* *External Channels*: Connect services across processes or instances through an external message broker, configured using MicroProfile Config.

*Channel Configuration*:

Channels are configured using MicroProfile Config, typically in `microprofile-config.properties` or as environment variables. Replace `<connector-name>` with the appropriate connector for your runtime, such as `liberty-kafka` for Open Liberty or `smallrye-kafka` for Quarkus:

[source,properties]
----
# Shared broker address for all channels

> mp.messaging.connector.liberty-kafka.bootstrap.servers=localhost:9092

Reviewer: If your examples use the liberty Kafka connector, I have no idea how it works, what's allowed and what is not.

Author: I have implemented and tested these examples as per documentation in this URL https://openliberty.io/docs/latest/liberty-kafka-connector.html.

mp.messaging.incoming.order-created.group.id=payment-service

# Outgoing channel to Kafka
mp.messaging.outgoing.payment-authorized.connector=liberty-kafka
mp.messaging.outgoing.payment-authorized.topic=payment-authorized-topic
----

*Channel Names*:

Channel names referenced in `@Incoming`, `@Outgoing`, and `@Channel` annotations must match the configuration property names:

[source,java]
----
@Incoming("order-created")  // Matches mp.messaging.incoming.order-created.*
@Outgoing("payment-authorized")
public Order processPayment(Order order) {
    // Processing logic
    return order;
}
----

Internal channels do not require connector configuration. For communication between separate services or multiple server instances, configure an external broker such as Kafka.

== Declarative Messaging with @Incoming and @Outgoing

Declarative messaging is at the heart of MicroProfile Reactive Messaging. By using simple annotations, developers can define complex message-processing pipelines without writing boilerplate code for message handling, threading, or backpressure management. The framework handles these concerns automatically based on the method signatures.

> In the MicroProfile e-commerce store application,the checkout begins with a synchronous REST request, and then downstream fulfillment steps are expressed declaratively through channels such as `order-created`, `payment-authorized`, and `inventory-reserved`.

Reviewer: ```suggestion
In the MicroProfile e-commerce store application, the checkout begins with a synchronous REST request, and then downstream fulfillment steps are expressed declaratively through channels such as `order-created`, `payment-authorized`, and `inventory-reserved`.
```

*Supported Method Signatures*:

*1. Simple Payload (Automatic Acknowledgment)*:

The runtime extracts the payload from the incoming message and passes it to the method. Acknowledgment is handled automatically: the message is acknowledged after the method returns normally, or negatively acknowledged if the method throws an exception. 

[source,java]
----
@Incoming("items")
public void process(String item) {
    // Process item
    // Message is automatically acknowledged after method returns
}
----

Use this form when your processing is synchronous and you do not need access to message metadata or manual acknowledgment control.

*2. Asynchronous Processing with CompletionStage*:

`java.util.concurrent.CompletionStage<T>` is used for asynchronous computation which may complete with a value or an exception. It can be chained using methods such as `thenApply`, `thenCompose`, and `exceptionally` to build non-blocking processing pipelines. 

[source,java]
----
@Incoming("items")
public CompletionStage<Void> processAsync(String item) {
    return processItemAsync(item);
    // Message acknowledged when CompletionStage completes
}
----

> When used as a method return type in MicroProfile Reactive Messaging, the message is ackno  ledged only after the `CompletionStage` completes successfully, giving you control over when acknowledgment happens relative to your async work. 

Reviewer: ```suggestion
When used as a method return type in MicroProfile Reactive Messaging, the message is acknowledged only after the `CompletionStage` completes successfully, giving you control over when acknowledgment happens relative to your async work. 
```

* `getPayload()` — returns the typed payload.
* `ack()` — returns a `CompletionStage<Void>` that positively acknowledges the message, signalling to the broker that processing succeeded.
* `nack(Throwable)` — negatively acknowledges the message, allowing the connector to retry or dead-letter it depending on its configuration.
* `getMetadata(Class<M>)` — retrieves optional connector-specific metadata such as topic, partition, or headers.

[source,java]
----
@Incoming("items")
public CompletionStage<Void> process(Message<String> message) {
    String payload = message.getPayload();

    return processItem(payload)
        .thenCompose(v -> message.ack())
        .exceptionally(throwable -> {
            message.nack(throwable);
            return null;
        });
}
----

Use `Message<T>` when you need explicit acknowledgment control, access to metadata, or custom error handling beyond what automatic acknowledgment provides.

*4. Reactive Streams Processing*:

> `Multi<T>` is a reactive type from the Mutiny library (`io.smallrye.mutiny.Multi`) that represents an asynchronous stream of zero or more items. It implements the Reactive Streams `Publisher<T>` contract and provides a rich set of operators for transforming, filtering, merging, and grouping streams in a composable, non-blocking way. Common operators include:

Reviewer: I think we need to distinguish things a bit.

The MP Reactive Messaging spec does not use Mutiny. It uses MP Reactive Streams Operator (which predates Mutiny, including types like `Publisher<T>`, `PublisherBuilder<T>`, `Subscriber<T>`, `SubscriberBuilder<T>`, `Processor<T,R>`, and `ProcessorBuilder<T,R>`) and CompletionStage from the JDK. Many of the code examples here use Mutiny. I think it should be clarified that while Mutiny can be used, it's one possibility among others (you can perfectly use Reactor Flux too). 

I let you decide if it's ok or not. I think it deserves a note.

Author: Done. I have added a note about this.

[source,java]
----
@Incoming("order-created")
@Outgoing("payment-authorized")
public Order processPayment(Order order) {
        LOGGER.info(() -> "Processing payment for order " + order.getOrderId());
        order.setStatus(OrderStatus.PAID);
        return order;
}
----

Each incoming `order-created` event produces exactly one outgoing message with the updated payment status.

*2. One-to-Many Transformation (Splitting)*:

Each incoming message produces multiple outgoing messages by returning a `Multi<T>`, allowing a single event to fan out into independent downstream items.

[source,java]
----
@Incoming("purchase-orders")
@Outgoing("line-items")
public Multi<LineItem> splitOrder(PurchaseOrder order) {
    return Multi.createFrom().iterable(order.getItems());
}
----

This code snippet above splits one incoming purchase order into multiple line-item messages, with each item sent downstream separately.

*3. Many-to-One Transformation (Aggregation)*:

> Multiple incoming messages are collected and combined into a single outgoing message, useful for batching, summarising, or reducing a stream of items into a consolidated result.

Reviewer: ```suggestion
Multiple incoming messages are collected and combined into a single outgoing message, useful for batching, summarizing, or reducing a stream of items into a consolidated result.
```

the rest of the chapter use US spelling.

*5. Asynchronous Enrichment*:

Each incoming payload triggers a non-blocking asynchronous operation, such as a database lookup, and the result is forwarded downstream only after the `CompletionStage` completes successfully.

[source,java]
----
@Incoming("user-ids")
@Outgoing("users")
public CompletionStage<User> enrichUser(String userId) {
    return userRepository.findById(userId);
}
----

In the above code snippet, each user ID triggers an asynchronous database lookup, and the enriched `User` object is sent downstream when the lookup completes.

*6. Conditional Routing (Dropping Messages)*:

Inspects each message's payload and either forwards it downstream or explicitly acknowledges and drops it by returning `null`, preventing unwanted messages from reaching the output channel.

[source,java]
----
@Incoming("transactions")
@Outgoing("high-value-transactions")
public Message<Transaction> routeHighValue(Message<Transaction> message) {
    Transaction tx = message.getPayload();

    if (tx.getAmount() > 10000) {
        return message;
    } else {
        message.ack();

>         return null;

Reviewer: This would work in SmallRye, but I'm not sure it is correct according to the spec.

* Use `CompletionStage<T>` for async operations like database lookups or REST calls.
* Use `Message<T>` when you need manual acknowledgment control or metadata access.
* Return `null` from `@Outgoing` methods to drop messages without forwarding.

== Programmatic Messaging with @Channel and @Emitter

In the MicroProfile e-commerce store, checkout still begins with a synchronous REST request. After the `order` service validates and persists the order, it can publish an `order-created` event so that payment, inventory, and shipment processing continue asynchronously. 

We use `@Channel` in the code for the direct access to a named stream, and use `Emitter` when imperative code such as a REST endpoint needs to publish into that stream.

=== Using @Channel to Access Channels

The `@Channel` annotation allows developers to inject a channel directly into a CDI bean for programmatic access to the message stream. This is useful when you want to observe a stream, adapt it to another reactive API, or integrate it with existing imperative code.

[source,java]
----
//..

@ApplicationScoped
public class PaymentEventMonitor {

    @Inject
    @Channel("payment-authorized")

>     Multi<Order> paymentEvents;

Reviewer: The support of Multi in @Channel is not in the spec, it's a SmallRye implementation feature. The spec only mentions `Publisher<T>` and `PublisherBuilder<T>`

*Sequential Processing Pattern*:

[source,java]
----
import org.eclipse.microprofile.reactive.messaging.Incoming;
import org.eclipse.microprofile.reactive.messaging.Outgoing;

@Incoming("data")
@Outgoing("processed-data")
public String process(String data) {
    return transform(data);
}
----

The framework automatically:

1. Requests one element from the upstream channel
2. Waits for the `process` method to complete
3. Sends the result downstream
4. Requests the next element only after the current one is fully processed

This sequential processing naturally provides backpressure: the upstream producer will not send more data until the current item is processed. If `transform(data)` takes 100ms, the consumer processes at most 10 messages per second, and the producer adapts to this rate.

*Asynchronous Processing with Backpressure*:

[source,java]
----
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CompletableFuture;

@Incoming("tasks")
@Outgoing("results")
public CompletionStage<Result> processAsync(Task task) {
    return CompletableFuture.supplyAsync(() -> 

>         expensiveOperation(task)

Reviewer: As I mentioned in another comment, this would use an unmanaged and uncontrolled F/J pool.

Configure buffer sizes for incoming channels to smooth out temporary speed mismatches:

[source,properties]
----
# Allow producer to send up to 256 messages ahead
mp.messaging.incoming.data.buffer-size=256
----

This creates a buffer of 256 messages between the connector and your consumer method. The producer can send messages at its natural rate while the consumer processes at its own pace, as long as the buffer doesn't fill completely.

*Choosing Buffer Size*:

* *Small buffers (10-50)*: Tight backpressure control, minimal memory usage, but frequent demand signals
* *Medium buffers (100-500)*: Good balance for most scenarios with occasional bursts
* *Large buffers (1000+)*: Tolerate significant bursts, but consume more memory and delay backpressure signals

*Concurrent Processing*:

Process multiple messages concurrently while maintaining overall backpressure control:

[source,java]
----
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;

@Incoming("tasks")
@Outgoing("results")
public Multi<Result> processConcurrent(Multi<Task> tasks) {
    return tasks
        .onItem().transformToUniAndMerge(task ->
            Uni.createFrom().completionStage(() ->

>                 CompletableFuture.supplyAsync(() -> process(task))

Reviewer: I would recommend NOT using supplyAsync without an explicit executor, but this goes beyond this book.

Author: Fixed.

            ),

>             3  // Process up to 3 messages concurrently

Reviewer: You cannot pass the concurrency level in `transformToUniAndMerge`, you need:

```
.onItem().transformToUni(task ->
    Uni.createFrom().completionStage(() ->
        CompletableFuture.supplyAsync(() -> process(task))
    )
).merge(3)
```

Author: Fixed.

Process messages in batches to improve throughput for operations that benefit from bulk processing:

[source,java]
----
import io.smallrye.mutiny.Multi;
import java.time.Duration;

@Incoming("events")
@Outgoing("batched-events")
public Multi<List<Event>> batch(Multi<Event> events) {
    return events
        .group().intoLists().of(100, Duration.ofSeconds(5));
}
----

This batches up to 100 events or waits up to 5 seconds, whichever comes first. Batching is particularly effective for:

* Bulk database inserts
* Batch API calls to external services
* Aggregated analytics computations

*Rate Limiting*:

Control the maximum processing rate to prevent overwhelming downstream systems:

[source,java]
----
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import java.time.Duration;

@Incoming("high-volume")
@Outgoing("rate-limited")
public Multi<Data> rateLimit(Multi<Data> stream) {
    return stream
        .group().intoLists().of(10)                    // Batch 10 items
        .onItem().call(batch ->                         // Pause between batches
            Uni.createFrom().nullItem()
                .onItem().delayIt().by(Duration.ofMillis(100))
        )

>         .onItem().transformToMultiAndConcatenate(Multi::createFrom().iterable);

Reviewer: This is not valid Java.

Reviewer: I believe you meant: `.onItem().transformToMultiAndConcatenate(batch -> Multi.createFrom().iterable(batch))
`

Author: > I believe you meant: `.onItem().transformToMultiAndConcatenate(batch -> Multi.createFrom().iterable(batch)) `

Yes, would changing it to `.onItem().transformToMultiAndConcatenate(Multi.createFrom()::iterable);` as batch being a std. `List` object, it will be implicitly passed by Java under the hood.

// Critical transactions that must all be processed
@Inject
@Channel("payments")
@OnOverflow(value = OnOverflow.Strategy.BUFFER, bufferSize = 1000)
Emitter<Payment> paymentEmitter;

// REST endpoint that should fail fast when overloaded
@Inject
@Channel("api-requests")
@OnOverflow(OnOverflow.Strategy.THROW_EXCEPTION)
Emitter<ApiRequest> requestEmitter;

@POST
@Path("/requests")
public Response handleRequest(ApiRequest request) {
    try {
        requestEmitter.send(request);
        return Response.accepted().build();
    } catch (IllegalStateException e) {
        return Response.status(503).entity("System overloaded").build();
    }
}
----

When the `@OnOverflow` annotation is absent, the `BUFFER` strategy with default buffer size applies automatically. This provides a sensible default for most applications while allowing fine-grained control when needed.

== Integration with Other MicroProfile Specifications

MicroProfile Reactive Messaging integrates seamlessly with other MicroProfile specifications, enabling comprehensive solutions for building resilient, observable, configurable, and secure microservices.

=== MicroProfile Fault Tolerance Integration

> MicroProfile Fault Tolerance complements reactive messaging by adding `@Retry`, `@Timeout`, `@CircuitBreaker`, `@Bulkhead`, and `@Fallback` to message-processing methods. This is covered in the `Message Acknowledgment and Error Handling` section. 

Reviewer: Did not find that section.

=== MicroProfile Metrics Integration

> When MicroProfile Reactive Messaging 3.0.1 is used in an environment where MicroProfile Metrics is available, implementations automatically produce the following metrics in the `base` scope:

Reviewer: Should it use the attribute instead of 3.0.1?

[cols="2,1,1,2"]
|===
|Metric Name |Type |Unit |Description

|`mp.messaging.message.count`
|Counter
|None
|The number of messages processed. Includes a `channel` tag identifying the channel name.
|===

Example metric output:
----
mp_messaging_message_count_total{channel="orders"} 1523
mp_messaging_message_count_total{channel="notifications"} 847
----

These metrics require no additional code or configuration and are automatically exposed via the MicroProfile Metrics endpoint (typically `/metrics`).

> === MicroProfile Health Integration

Reviewer: that should be extended.

> MicroProfile Health enables startup, liveness and readiness checkes for messaging components, essential for container orchestration platforms like Kubernetes.

Reviewer: ```suggestion
MicroProfile Health enables startup, liveness, and readiness checks for messaging components, essential for container orchestration platforms like Kubernetes.
```

=== MicroProfile Config Integration

MicroProfile Config enables externalized configuration for messaging channels, connectors, and application properties, supporting multiple configuration sources with defined precedence.

> == Reactive Stream Operations

Reviewer: This section is a bit limited...

Reactive stream operations provide powerful tools for transforming, filtering, and combining message streams. MicroProfile Reactive Messaging implementations support reactive operators through libraries like Mutiny, enabling functional-style stream processing with backpressure management.

Reactive stream operations enable building sophisticated message processing pipelines declaratively while maintaining backpressure and resource efficiency throughout the entire flow.

== Graph Validation and Deployment Checks

When the application starts, the MicroProfile Reactive Messaging implementation constructs a directed graph connecting all `@Incoming` and `@Outgoing` methods, `Emitter` instances, `@Channel` injections, and connectors. The implementation validates this graph and throws a `DeploymentException` if any of the following invalid conditions exist:

*Invalid Channel Configurations*:

* A method annotated with `@Incoming` has no upstream channel
* A method annotated with `@Outgoing` has no downstream channel

> * A method annotated with `@Incoming` has multiple upstream channels (fan-in not supported)

Reviewer: not in the spec, SmallRye does, using @Merge.

Reviewer: I've found the wording a bit weird.

> * A method annotated with `@Outgoing` has multiple downstream channels (fan-out requires multiple methods)

Reviewer: not in the spec, but SmallRye does.

The declarative messaging model with `@Incoming` and `@Outgoing` annotations eliminates boilerplate code for message routing and transformation, allowing developers to focus on business logic rather than infrastructure concerns. When declarative patterns are insufficient, `@Channel` and `@Emitter` provide imperative message sending capabilities, particularly useful for bridging REST endpoints to message streams. The framework handles backpressure automatically through reactive operators, configurable buffer sizes, and `@OnOverflow` strategies, with `BUFFER` defaulting to 128 elements to prevent overwhelming downstream consumers.

> Reliable message processing is ensured through multiple acknowledgment strategies including `POST_PROCESSING`, `PRE_PROCESSING`, `MANUAL`, and `NONE`, along with comprehensive error handling patterns that support dead letter queues and retry policies. The runtime performs graph validation at startup, throwing `DeploymentException` for misconfigured channels to ensure failures are detected before production deployment rather than at runtime. When MicroProfile Metrics is available, the implementation automatically produces metrics per channel such as `mp.messaging.message.count`, complemented by integration with MicroProfile Health for liveness and readiness probes.

Reviewer: This has not been explained.

Also dead letter queue and retries depends on the connector you use.

The specification's seamless integration with other MicroProfile specifications enhances the overall capabilities of reactive messaging applications. MicroProfile Fault Tolerance provides retry, circuit breaker, and timeout patterns for resilience; Config enables externalized configuration for different environments; Metrics supports both automatic and custom observability; Health integrates with Kubernetes for readiness probes; and OpenAPI documents REST endpoints that bridge to messaging systems. The connector ecosystem, supporting Kafka, AMQP, JMS, and other brokers, handles protocol-specific details, serialization, and connection management, freeing applications to focus purely on message processing logic.

> To continue your journey with MicroProfile Reactive Messaging, explore the complete https://download.eclipse.org/microprofile/microprofile-reactive-messaging-3.0.1/microprofile-reactive-messaging-spec-3.0.1.html[MicroProfile Reactive Messaging 3.0.1 Specification] for detailed API documentation and conformance requirements. Study connector-specific features for your chosen message broker in your runtime's documentation. Leverage the integration points with other MicroProfile specifications such as Fault Tolerance for resilience, Config for externalization, Metrics for observability, and Health for container orchestration.

Reviewer: Should it use the attribute instead of 3.0.1?

Author: Done.

By mastering these concepts and patterns, you can build modern, reactive microservices that handle high-volume data streams efficiently and reliably in production environments, taking full advantage of the reactive programming model to create responsive, resilient, and scalable distributed systems.

> This concludes the MicroProfile tutorial. You are now equipped with the foundational knowledge to build robust, cloud-native microservices using the MicroProfile specification. Thank you for following along, and happy coding!

Reviewer: ```suggestion
This concludes the chapter on MicroProfile Reactive Messaging. You are now equipped with the foundational knowledge to build robust, cloud-native microservices using the MicroProfile specification. Thank you for following along, and happy coding!
```
